time.c

```diff
@@ -5215,6 +5215,78 @@ time_strftime(VALUE time, VALUE format)
     }
 }
 
+static VALUE
+time_xmlschema(int argc, VALUE *argv, VALUE time)
+{
+    long fraction_digits = 0;
+    rb_check_arity(argc, 0, 1);
+    if (argc > 0) {
+        fraction_digits = NUM2LONG(argv[0]);
+        if (fraction_digits < 0) {
+            fraction_digits = 0;
+        }
+    }
+
+    struct time_object *tobj;
+
+    GetTimeval(time, tobj);
+    MAKE_TM(time, tobj);
+
+    long year = -1;
+    if (FIXNUM_P(tobj->vtm.year)) {
+        year = FIX2LONG(tobj->vtm.year);
+    }
+    if (RB_UNLIKELY(year > 9999 || year < 0 || fraction_digits > 9)) {
+        // Slow path for uncommon dates.
+        VALUE format = rb_utf8_str_new_cstr("%FT%T");
+        if (fraction_digits > 0) {
+            rb_str_catf(format, ".%%#%ldN", fraction_digits);
+        }
+        rb_str_cat_cstr(format, TZMODE_UTC_P(tobj) ? "Z" : "%:z");
+        return rb_funcallv(time, rb_intern("strftime"), 1, &format);
+    }
+
+    long buf_size = sizeof("YYYY-MM-DDTHH:MM:SS+ZH:ZM") + fraction_digits + (fraction_digits > 0 ? 1 : 0);
+
+    VALUE str = rb_str_buf_new(buf_size);
+    rb_enc_associate_index(str, rb_utf8_encindex());
+
+    char *ptr = RSTRING_PTR(str);
+    char *start = ptr;
+    int written = snprintf(ptr, sizeof("YYYY-MM-DDTHH:MM:SS"), "%04ld-%02d-%02dT%02d:%02d:%02d", year, tobj->vtm.mon, tobj->vtm.mday, tobj->vtm.hour, tobj->vtm.min, tobj->vtm.sec);
+    RUBY_ASSERT(written == sizeof("YYYY-MM-DDTHH:MM:SS"));
+    ptr += written;
+
+    if (fraction_digits > 0) {
+        long nsec = NUM2LONG(mulquov(tobj->vtm.subsecx, INT2FIX(1000000000), INT2FIX(TIME_SCALE)));
+        long subsec = nsec / powl(10, 9 - fraction_digits);
```

Comment:
`double` is enough for 8-digits number.
Rather FP-division may result in repeating binary.

```suggestion
        long subsec = nsec / (long)pow(10, 9 - fraction_digits);
```